dotcom-rendering/src/model/insert-newsletter-signup.ts

```diff
@@ -0,0 +1,167 @@
+	elements[index - 1]?._type ===
+	'model.dotcomrendering.pageElements.TextBlockElement';
+
+const getPlaces = (
+	targetIndex: number,
+	elements: CAPIElement[],
+): PlaceInArticle[] => {
+	return elements.map((element, index) => ({
+		index,
+		distance: Math.abs(index - targetIndex),
+		afterText: isAfterText(index, elements),
+		afterSupporting: isCloseAfterSupportingElement(index, elements),
+	}));
+};
+
+const placeIsSuitable = (place: PlaceInArticle): boolean =>
+	place.afterText && !place.afterSupporting;
+
+/**
+ * Finds a place within the existing element of the block to insert a
+ * NewsletterSignupBlockElement, within a given distance from a given
+ * target index, prefering a place after TextBlockElement and not within
+ * two places after a supporting element (eg pullquote).
+ *
+ * @param elements the elements in the article block
+ * @param targetIndex the target place to put the NewsletterSignupBlockElement
+ * @param maxDistance how far to move from the targetIndex to find a place after a paragraph
+ * @returns the index insert the NewsletterSignupBlockElement component at
+ */
+export const findInsertIndex = (
+	elements: CAPIElement[],
+	targetIndex: number,
+	maxDistance = 0,
+): number => {
+	const possiblePlaces = getPlaces(targetIndex, elements)
+		.filter(placeIsSuitable)
+		.filter((place) => place.distance <= maxDistance)
+		.sort((a, b) => a.distance - b.distance); // sort, closest place first
+
+	// Return index of the closest place - if there are none, return the targetIndex
+	return possiblePlaces[0]?.index || targetIndex;
+};
+
+/**
+ * Determines whether the element should have clear:both
+ * for all breakpoints
+ *
+ * @param index where in the array the element will be inserted
+ * @param elements the array of elements
+ * @returns whether the element should have clear:both for all breakpoints
+ */
+const decideIfClearFromLeftCol = (
+	index: number,
+	elements: CAPIElement[],
+): boolean => {
+	const floatingElementRoleTypes = ['supporting', 'thumbnail', 'richLink'];
+
+	const lastFloatingElementBeforePlace = elements
+		.slice(0, index)
+		.reverse()
+		.find(
+			(element) =>
+				'role' in element &&
+				typeof element.role == 'string' &&
+				floatingElementRoleTypes.includes(element.role),
+		);
+
+	// Ideally wouldn't need to test for "'role' in element" again
+	const lastFloatingElementRole =
+		lastFloatingElementBeforePlace &&
+		'role' in lastFloatingElementBeforePlace
+			? lastFloatingElementBeforePlace.role
+			: undefined;
+
+	return lastFloatingElementRole === 'supporting';
+};
+
+// NOTE: blog pages have different structure with multiple blocks of elements
+// any future function to place sign-up blocks in blog pages would need to
+// take this into acccount.
+const insert = (
+	newsletter: Newsletter,
+	elements: CAPIElement[],
+	targetIndex: number,
+	maxDistance = 0,
+): CAPIElement[] => {
+	const index = findInsertIndex(elements, targetIndex, maxDistance);
+
+	elements.splice(index, 0, {
```

Comment:
`splice` will mutate the original array, which can potentially cause hard to reason about issues. Is there a solution here using destructuring?

Reply:
good point - I've changed this to `[...array.slice(0, index), newItem, ...array.slice(index)]`